File: clients/web/src/modules/search/services/queryParser.ts
/**
 * Query Parser Service
 * Parses search queries into structured form with filters, phrases, and intent detection
 */

import type { ModuleType } from '@/types/modules';
import type {
  ParsedQuery,
  QueryFilter,
  QueryIntent,
  SearchScope,
} from '../types';

// ============================================================================
// Constants
// ============================================================================

/**
 * Organizing-specific concept expansions
 * Maps common organizing terms to related concepts
 */
const CONCEPT_EXPANSIONS: Record<string, string[]> = {
  // Events
  meeting: ['event', 'gathering', 'session', 'call', 'assembly', 'caucus'],
  event: ['meeting', 'gathering', 'action', 'rally', 'march'],
  rally: ['march', 'protest', 'demonstration', 'action', 'event'],
  action: ['event', 'rally', 'protest', 'campaign', 'direct action'],

  // Finance
  budget: ['finance', 'money', 'funding', 'costs', 'expenses', 'dues'],
  dues: ['membership', 'payment', 'contribution', 'budget', 'finance'],
  fundraising: ['donation', 'campaign', 'money', 'grant', 'crowdfunding'],

  // Governance
  vote: ['ballot', 'poll', 'election', 'proposal', 'decision'],
  proposal: ['motion', 'resolution', 'vote', 'decision', 'initiative'],
  election: ['vote', 'ballot', 'candidate', 'nomination'],

  // Organizing
  campaign: ['drive', 'effort', 'initiative', 'action', 'movement'],
  outreach: ['canvassing', 'phonebank', 'doorknock', 'tabling'],
  canvassing: ['doorknock', 'outreach', 'petitioning', 'signatures'],

  // Communication
  message: ['dm', 'chat', 'conversation', 'thread'],
  announcement: ['news', 'update', 'notice', 'broadcast'],

  // Mutual aid
  request: ['need', 'ask', 'help', 'assistance'],
  offer: ['give', 'provide', 'volunteer', 'share'],
  rideshare: ['carpool', 'ride', 'transport', 'pickup'],

  // Documentation
  document: ['doc', 'file', 'paper', 'record'],
  wiki: ['knowledge', 'documentation', 'guide', 'reference'],
  minutes: ['notes', 'record', 'transcript', 'summary'],
};

/**
 * Common stop words to filter out
 */
const STOP_WORDS = new Set([
  'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for',
  'from', 'has', 'he', 'in', 'is', 'it', 'its', 'of', 'on',
  'or', 'she', 'that', 'the', 'they', 'to', 'was', 'were',
  'which', 'with', 'you', 'your', 'this', 'these', 'those',
]);

/**
 * Filter operators for query syntax
 */
const FILTER_PATTERNS: Array<{
  pattern: RegExp;
  field: string;
  operator: QueryFilter['operator'];
  transform?: (match: string) => unknown;
}> = [
  // @author:pubkey or @username
  {
    pattern: /@(\w+)/g,
    field: 'author',
    operator: 'eq',
    transform: (match) => match,
  },
  // #tag
  {
    pattern: /#([\w-]+)/g,
    field: 'tag',
    operator: 'contains',
    transform: (match) => match.toLowerCase(),
  },
  // type:events or module:events
  {
    pattern: /(?:type|module):(\w+)/gi,
    field: 'moduleType',
    operator: 'eq',
    transform: (match) => match.toLowerCase() as ModuleType,
  },
  // group:groupId
  {
    pattern: /group:([\w-]+)/gi,
    field: 'groupId',
    operator: 'eq',
    transform: (match) => match,
  },
  // date:2024-01-01 or date:today or date:yesterday
  {
    pattern: /date:([\w-]+)/gi,
    field: 'date',
    operator: 'eq',
    transform: (match) => parseDateValue(match),
  },
  // before:2024-01-01
  {
    pattern: /before:([\w-]+)/gi,
    field: 'date',
    operator: 'lt',
    transform: (match) => parseDateValue(match),
  },
  // after:2024-01-01
  {
    pattern: /after:([\w-]+)/gi,
    field: 'date',
    operator: 'gt',
    transform: (match) => parseDateValue(match),
  },
  // status:open or status:closed
  {
    pattern: /status:(\w+)/gi,
    field: 'status',
    operator: 'eq',
    transform: (match) => match.toLowerCase(),
  },
];

/**
 * Temporal patterns for intent detection
 */
const TEMPORAL_PATTERNS: Array<{
  pattern: RegExp;
  intent: QueryIntent;
}> = [
  {
    pattern: /\b(upcoming|future|next|soon)\b/i,
    intent: {
      type: 'temporal',
      params: { direction: 'future', relative: true },
    },
  },
  {
    pattern: /\b(past|previous|last|recent|earlier)\b/i,
    intent: {
      type: 'temporal',
      params: { direction: 'past', relative: true },
    },
  },
  {
    pattern: /\b(today|tonight)\b/i,
    intent: {
      type: 'temporal',
      params: { period: 'today' },
    },
  },
  {
    pattern: /\b(tomorrow)\b/i,
    intent: {
      type: 'temporal',
      params: { period: 'tomorrow' },
    },
  },
  {
    pattern: /\b(yesterday)\b/i,
    intent: {
      type: 'temporal',
      params: { period: 'yesterday' },
    },
  },
  {
    pattern: /\bthis\s+(week|month|year)\b/i,
    intent: {
      type: 'temporal',
      params: { period: 'this', unit: '$1' },
    },
  },
  {
    pattern: /\blast\s+(week|month|year)\b/i,
    intent: {
      type: 'temporal',
      params: { period: 'last', unit: '$1' },
    },
  },
];

// ============================================================================
// Parser Functions
// ============================================================================

/**
 * Parse a date value from query syntax
 */
function parseDateValue(value: string): number {
  const now = new Date();
  const lowered = value.toLowerCase();

  switch (lowered) {
    case 'today':
      return startOfDay(now).getTime();
    case 'yesterday':
      return startOfDay(new Date(now.getTime() - 24 * 60 * 60 * 1000)).getTime();
    case 'tomorrow':
      return startOfDay(new Date(now.getTime() + 24 * 60 * 60 * 1000)).getTime();
    case 'thisweek':
    case 'this-week':
      return startOfWeek(now).getTime();
    case 'lastweek':
    case 'last-week':
      return startOfWeek(new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000)).getTime();
    case 'thismonth':
    case 'this-month':
      return new Date(now.getFullYear(), now.getMonth(), 1).getTime();
    default:
      // Try to parse as ISO date
      const parsed = Date.parse(value);
      return isNaN(parsed) ? now.getTime() : parsed;
  }
}

function startOfDay(date: Date): Date {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
}

function startOfWeek(date: Date): Date {
  const d = new Date(date);
  const day = d.getDay();
  const diff = d.getDate() - day;
  return new Date(d.setDate(diff));
}

/**
 * Extract quoted phrases from query
 */
function extractPhrases(query: string): { phrases: string[]; remaining: string } {
  const phrases: string[] = [];
  let remaining = query;

  // Match "quoted phrases" or 'quoted phrases'
  const phrasePattern = /["']([^"']+)["']/g;
  let match;

  while ((match = phrasePattern.exec(query)) !== null) {
    phrases.push(match[1].trim());
    remaining = remaining.replace(match[0], ' ');
  }

  return { phrases, remaining: remaining.trim() };
}

/**
 * Extract filters from query syntax
 */
function extractFilters(query: string): { filters: QueryFilter[]; remaining: string } {
  const filters: QueryFilter[] = [];
  let remaining = query;

  for (const { pattern, field, operator, transform } of FILTER_PATTERNS) {
    // Reset lastIndex for global patterns
    pattern.lastIndex = 0;
    let match;

    while ((match = pattern.exec(query)) !== null) {
      const rawValue = match[1];
      const value = transform ? transform(rawValue) : rawValue;

      filters.push({
        field,
        operator,
        value: value as string | number | boolean | string[],
      });

      remaining = remaining.replace(match[0], ' ');
    }
  }

  return { filters, remaining: remaining.trim() };
}

/**
 * Detect query intent from natural language patterns
 */
function detectIntent(query: string): QueryIntent | undefined {
  for (const { pattern, intent } of TEMPORAL_PATTERNS) {
    const match = query.match(pattern);
    if (match) {
      // Replace placeholders in params with captured groups
      const params = { ...intent.params };
      if (match[1] && params.unit === '$1') {
        params.unit = match[1].toLowerCase();
      }
      return { ...intent, params };
    }
  }

  return undefined;
}

/**
 * Tokenize and stem keywords
 */
function tokenizeKeywords(text: string): string[] {
  // Split on whitespace and punctuation
  const tokens = text
    .toLowerCase()
    .split(/[\s\-_.,;:!?()[\]{}'"]+/)
    .filter(Boolean);

  // Filter stop words and short tokens
  return tokens.filter(
    (token) => token.length > 1 && !STOP_WORDS.has(token)
  );
}

/**
 * Expand keywords with synonyms/related concepts
 */
function expandKeywords(keywords: string[]): string[] {
  const expanded = new Set<string>();

  for (const keyword of keywords) {
    expanded.add(keyword);

    // Check for concept expansions
    const synonyms = CONCEPT_EXPANSIONS[keyword];
    if (synonyms) {
      for (const syn of synonyms) {
        expanded.add(syn);
      }
    }
  }

  return Array.from(expanded);
}

/**
 * Simple stemmer for English words
 * Handles common suffixes
 */
function stem(word: string): string {
  // Very basic stemming - could use a proper library for production
  return word
    .replace(/ies$/, 'y')
    .replace(/es$/, '')
    .replace(/s$/, '')
    .replace(/ing$/, '')
    .replace(/ed$/, '')
    .replace(/ly$/, '')
    .replace(/ment$/, '')
    .replace(/tion$/, '')
    .replace(/ness$/, '');
}

// ============================================================================
// Main Parser
// ============================================================================

/**
 * Parse a raw search query into a structured ParsedQuery
 */
export function parseQuery(raw: string, scope: SearchScope): ParsedQuery {
  if (!raw || typeof raw !== 'string') {
    return {
      raw: '',
      keywords: [],
      phrases: [],
      expandedTerms: [],
      filters: [],
      intent: undefined,
      scope,
    };
  }

  const trimmed = raw.trim();

  // Extract quoted phrases first
  const { phrases, remaining: afterPhrases } = extractPhrases(trimmed);

  // Extract filter syntax
  const { filters, remaining: afterFilters } = extractFilters(afterPhrases);

  // Detect intent from remaining text
  const intent = detectIntent(afterFilters);

  // Tokenize remaining text into keywords
  const keywords = tokenizeKeywords(afterFilters);

  // Stem keywords
  const stemmedKeywords = keywords.map(stem);

  // Expand with synonyms
  const expandedTerms = expandKeywords(stemmedKeywords);

  return {
    raw: trimmed,
    keywords: stemmedKeywords,
    phrases,
    expandedTerms,
    filters,
    intent,
    scope,
  };
}

/**
 * Build a display-friendly representation of the parsed query
 */
export function formatParsedQuery(query: ParsedQuery): string {
  const parts: string[] = [];

  if (query.phrases.length > 0) {
    parts.push(query.phrases.map((p) => `"${p}"`).join(' '));
  }

  if (query.keywords.length > 0) {
    parts.push(query.keywords.join(' '));
  }

  if (query.filters.length > 0) {
    const filterStrs = query.filters.map((f) => {
      switch (f.field) {
        case 'author':
          return `@${f.value}`;
        case 'tag':
          return `#${f.value}`;
        default:
          return `${f.field}:${f.value}`;
      }
    });
    parts.push(filterStrs.join(' '));
  }

  return parts.join(' ') || query.raw;
}

/**
 * Check if a query has any meaningful content
 */
export function isEmptyQuery(query: ParsedQuery): boolean {
  return (
    query.keywords.length === 0 &&
    query.phrases.length === 0 &&
    query.filters.length === 0
  );
}

/**
 * Get concept expansions dictionary
 */
export function getConceptExpansions(): Record<string, string[]> {
  return { ...CONCEPT_EXPANSIONS };
}

/**
 * Add custom concept expansion
 */
export function addConceptExpansion(term: string, synonyms: string[]): void {
  CONCEPT_EXPANSIONS[term.toLowerCase()] = synonyms.map((s) => s.toLowerCase());
}

File: clients/ios/BuildIt/Modules/Search/Services/QueryParser.swift
// QueryParser.swift
// BuildIt - Decentralized Mesh Communication
//
// Parses search query syntax including phrases, filters, and operators.
// Supports organizing-focused concept expansion.

import Foundation
import os.log

// MARK: - QueryParser

/// Parses and normalizes search queries
public struct QueryParser: Sendable {
    // MARK: - Properties

    private let logger = Logger(subsystem: "com.buildit", category: "QueryParser")
    private let conceptExpansions: [String: ConceptExpansion]

    // MARK: - Initialization

    public init() {
        // Initialize with organizing-focused concept expansions
        self.conceptExpansions = Self.defaultConceptExpansions
    }

    // MARK: - Public API

    /// Parse a raw query string into a structured ParsedQuery
    /// - Parameters:
    ///   - query: The raw query string
    ///   - scope: The search scope
    /// - Returns: A parsed and normalized query
    public func parse(_ query: String, scope: SearchScope = .global) -> ParsedQuery {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)

        // Extract quoted phrases
        let (withoutPhrases, phrases) = extractPhrases(from: trimmed)

        // Extract filters (e.g., author:name, tag:value, date:2024-01)
        let (withoutFilters, filters) = extractFilters(from: withoutPhrases)

        // Extract keywords
        let keywords = extractKeywords(from: withoutFilters)

        // Expand concepts for semantic search
        let expandedTerms = expandConcepts(keywords: keywords)

        return ParsedQuery(
            raw: query,
            keywords: keywords,
            phrases: phrases,
            expandedTerms: expandedTerms.isEmpty ? nil : expandedTerms,
            filters: filters,
            scope: scope
        )
    }

    /// Stem a word for matching
    public func stem(_ word: String) -> String {
        // Simple English stemmer - in production, use a proper stemmer library
        var stemmed = word.lowercased()

        // Remove common suffixes
        let suffixes = ["ing", "ed", "ly", "es", "s", "er", "est", "ment", "ness", "tion", "sion"]
        for suffix in suffixes {
            if stemmed.hasSuffix(suffix) && stemmed.count > suffix.count + 2 {
                stemmed = String(stemmed.dropLast(suffix.count))
                break
            }
        }

        return stemmed
    }

    /// Check if a query is empty or only contains operators
    public func isEmpty(_ query: String) -> Bool {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let parsed = parse(trimmed)
        return parsed.keywords.isEmpty && parsed.phrases.isEmpty && parsed.filters.isEmpty
    }

    // MARK: - Private Methods

    /// Extract quoted phrases from query
    private func extractPhrases(from query: String) -> (remaining: String, phrases: [String]) {
        var remaining = query
        var phrases: [String] = []

        // Match double-quoted phrases
        let doubleQuotePattern = "\"([^\"]+)\""
        if let regex = try? NSRegularExpression(pattern: doubleQuotePattern) {
            let range = NSRange(remaining.startIndex..., in: remaining)
            let matches = regex.matches(in: remaining, range: range)

            for match in matches.reversed() {
                if let phraseRange = Range(match.range(at: 1), in: remaining) {
                    phrases.insert(String(remaining[phraseRange]), at: 0)
                }
                if let fullRange = Range(match.range, in: remaining) {
                    remaining.removeSubrange(fullRange)
                }
            }
        }

        // Match single-quoted phrases
        let singleQuotePattern = "'([^']+)'"
        if let regex = try? NSRegularExpression(pattern: singleQuotePattern) {
            let range = NSRange(remaining.startIndex..., in: remaining)
            let matches = regex.matches(in: remaining, range: range)

            for match in matches.reversed() {
                if let phraseRange = Range(match.range(at: 1), in: remaining) {
                    phrases.insert(String(remaining[phraseRange]), at: 0)
                }
                if let fullRange = Range(match.range, in: remaining) {
                    remaining.removeSubrange(fullRange)
                }
            }
        }

        return (remaining, phrases)
    }

    /// Extract filter expressions from query
    private func extractFilters(from query: String) -> (remaining: String, filters: [QueryFilter]) {
        var remaining = query
        var filters: [QueryFilter] = []

        // Pattern for filters like field:value or field:"quoted value"
        let filterPattern = "(author|tag|type|module|group|from|to|date|before|after):(?:\"([^\"]+)\"|([^\\s]+))"

        if let regex = try? NSRegularExpression(pattern: filterPattern, options: .caseInsensitive) {
            let range = NSRange(remaining.startIndex..., in: remaining)
            let matches = regex.matches(in: remaining, range: range)

            for match in matches.reversed() {
                guard let fieldRange = Range(match.range(at: 1), in: remaining) else { continue }
                let field = String(remaining[fieldRange]).lowercased()

                // Get value from either quoted or unquoted capture group
                var value: String?
                if match.range(at: 2).location != NSNotFound,
                   let valueRange = Range(match.range(at: 2), in: remaining) {
                    value = String(remaining[valueRange])
                } else if let valueRange = Range(match.range(at: 3), in: remaining) {
                    value = String(remaining[valueRange])
                }

                if let value = value {
                    if let filter = createFilter(field: field, value: value) {
                        filters.insert(filter, at: 0)
                    }
                }

                if let fullRange = Range(match.range, in: remaining) {
                    remaining.removeSubrange(fullRange)
                }
            }
        }

        return (remaining, filters)
    }

    /// Create a filter from field and value
    private func createFilter(field: String, value: String) -> QueryFilter? {
        switch field {
        case "author", "from":
            return QueryFilter(field: "authorPubkey", operator: .contains, value: .string(value))

        case "tag":
            return QueryFilter(field: "tags", operator: .contains, value: .string(value))

        case "type", "module":
            return QueryFilter(field: "moduleType", operator: .eq, value: .string(value))

        case "group":
            return QueryFilter(field: "groupId", operator: .eq, value: .string(value))

        case "date":
            // Parse date value
            if let timestamp = parseDate(value) {
                let dayStart = timestamp
                let dayEnd = timestamp + 86400000 // +24 hours in ms
                return QueryFilter(field: "createdAt", operator: .range, value: .range(min: Double(dayStart), max: Double(dayEnd)))
            }

        case "before":
            if let timestamp = parseDate(value) {
                return QueryFilter(field: "createdAt", operator: .lt, value: .number(Double(timestamp)))
            }

        case "after":
            if let timestamp = parseDate(value) {
                return QueryFilter(field: "createdAt", operator: .gt, value: .number(Double(timestamp)))
            }

        default:
            break
        }

        return nil
    }

    /// Parse a date string to Unix timestamp (ms)
    private func parseDate(_ value: String) -> Int64? {
        let formatters: [DateFormatter] = {
            let formats = ["yyyy-MM-dd", "yyyy-MM", "yyyy", "MM/dd/yyyy", "dd/MM/yyyy"]
            return formats.map { format in
                let formatter = DateFormatter()
                formatter.dateFormat = format
                return formatter
            }
        }()

        for formatter in formatters {
            if let date = formatter.date(from: value) {
                return Int64(date.timeIntervalSince1970 * 1000)
            }
        }

        // Try relative dates
        let lowercased = value.lowercased()
        let now = Date()
        let calendar = Calendar.current

        switch lowercased {
        case "today":
            return Int64(calendar.startOfDay(for: now).timeIntervalSince1970 * 1000)
        case "yesterday":
            if let date = calendar.date(byAdding: .day, value: -1, to: now) {
                return Int64(calendar.startOfDay(for: date).timeIntervalSince1970 * 1000)
            }
        case "thisweek":
            if let date = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: now)) {
                return Int64(date.timeIntervalSince1970 * 1000)
            }
        case "thismonth":
            if let date = calendar.date(from: calendar.dateComponents([.year, .month], from: now)) {
                return Int64(date.timeIntervalSince1970 * 1000)
            }
        default:
            break
        }

        return nil
    }

    /// Extract keywords from query
    private func extractKeywords(from query: String) -> [String] {
        // Remove special characters and split by whitespace
        let cleaned = query.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty && $0.count >= 2 }

        // Remove stop words
        let stopWords: Set<String> = [
            "the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for",
            "of", "with", "by", "from", "as", "is", "was", "are", "were", "been",
            "be", "have", "has", "had", "do", "does", "did", "will", "would",
            "could", "should", "may", "might", "must", "shall", "can", "need",
            "that", "this", "these", "those", "it", "its", "they", "them", "their",
            "we", "us", "our", "you", "your", "he", "she", "him", "her", "his"
        ]

        return cleaned.filter { !stopWords.contains($0) }
    }

    /// Expand concepts with synonyms and related terms
    private func expandConcepts(keywords: [String]) -> [String] {
        var expanded: Set<String> = []

        for keyword in keywords {
            let stemmed = stem(keyword)

            // Check direct expansions
            if let expansion = conceptExpansions[stemmed] {
                expanded.formUnion(expansion.synonyms)
                if let narrower = expansion.narrower {
                    expanded.formUnion(narrower)
                }
            }

            // Check if keyword matches any synonym
            for (term, expansion) in conceptExpansions {
                if expansion.synonyms.contains(stemmed) {
                    expanded.insert(term)
                    expanded.formUnion(expansion.synonyms)
                }
            }
        }

        // Remove original keywords from expansion
        let keywordSet = Set(keywords.map { stem($0) })
        return Array(expanded.subtracting(keywordSet))
    }

    // MARK: - Concept Expansions

    /// Default concept expansions for organizing contexts
    private static let defaultConceptExpansions: [String: ConceptExpansion] = [
        "meeting": ConceptExpansion(
            term: "meeting",
            synonyms: ["assembly", "gathering", "session", "conference", "huddle", "standup"],
            broader: "event",
            narrower: ["general assembly", "committee meeting", "caucus"]
        ),
        "vote": ConceptExpansion(
            term: "vote",
            synonyms: ["ballot", "poll", "election", "referendum", "decision"],
            broader: "governance",
            narrower: ["motion", "resolution", "amendment"]
        ),
        "action": ConceptExpansion(
            term: "action",
            synonyms: ["campaign", "protest", "rally", "demonstration", "march", "strike", "boycott"],
            broader: "organizing",
            narrower: ["direct action", "civil disobedience", "sit-in"]
        ),
        "member": ConceptExpansion(
            term: "member",
            synonyms: ["participant", "organizer", "volunteer", "activist", "comrade", "ally"],
            broader: "person"
        ),
        "mutual aid": ConceptExpansion(
            term: "mutual aid",
            synonyms: ["solidarity", "support", "assistance", "help", "resource sharing"],
            broader: "organizing"
        ),
        "union": ConceptExpansion(
            term: "union",
            synonyms: ["labor union", "trade union", "workers union", "guild", "collective"],
            broader: "organization",
            narrower: ["local", "chapter", "branch"]
        ),
        "coop": ConceptExpansion(
            term: "coop",
            synonyms: ["cooperative", "co-op", "worker cooperative", "collective", "commune"],
            broader: "organization"
        ),
        "proposal": ConceptExpansion(
            term: "proposal",
            synonyms: ["motion", "resolution", "recommendation", "suggestion", "initiative"],
            broader: "governance"
        ),
        "consensus": ConceptExpansion(
            term: "consensus",
            synonyms: ["agreement", "accord", "unity", "collective decision"],
            broader: "governance"
        ),
        "fundrais": ConceptExpansion(
            term: "fundraising",
            synonyms: ["donation", "contribution", "crowdfunding", "campaign", "drive"],
            broader: "resources"
        ),
        "outreach": ConceptExpansion(
            term: "outreach",
            synonyms: ["canvassing", "door-knocking", "phone banking", "tabling", "recruitment"],
            broader: "organizing"
        ),
        "train": ConceptExpansion(
            term: "training",
            synonyms: ["workshop", "education", "skill-building", "orientation", "onboarding"],
            broader: "event"
        ),
        "document": ConceptExpansion(
            term: "document",
            synonyms: ["file", "record", "report", "minutes", "notes", "resource"],
            broader: nil
        )
    ]
}

// MARK: - Query Suggestion

/// Suggests search queries based on partial input
public struct QuerySuggester: Sendable {
    private let recentSearches: [RecentSearch]
    private let savedSearches: [SavedSearch]
    private let popularTerms: [String]

    public init(
        recentSearches: [RecentSearch] = [],
        savedSearches: [SavedSearch] = [],
        popularTerms: [String] = []
    ) {
        self.recentSearches = recentSearches
        self.savedSearches = savedSearches
        self.popularTerms = popularTerms
    }

    /// Get suggestions for a partial query
    /// - Parameter prefix: The partial query string
    /// - Returns: Array of suggested queries
    public func suggest(for prefix: String) -> [String] {
        let lowercased = prefix.lowercased()
        var suggestions: [String] = []

        // Add matching recent searches
        for recent in recentSearches where recent.query.lowercased().hasPrefix(lowercased) {
            if !suggestions.contains(recent.query) {
                suggestions.append(recent.query)
            }
        }

        // Add matching saved searches
        for saved in savedSearches where saved.query.lowercased().hasPrefix(lowercased) {
            if !suggestions.contains(saved.query) {
                suggestions.append(saved.query)
            }
        }

        // Add filter suggestions if typing a filter
        if lowercased.contains(":") {
            suggestions.append(contentsOf: suggestFilters(for: lowercased))
        }

        // Limit suggestions
        return Array(suggestions.prefix(10))
    }

    /// Suggest filter completions
    private func suggestFilters(for prefix: String) -> [String] {
        let filterPrefixes = [
            "author:", "tag:", "type:", "module:", "group:",
            "date:", "before:", "after:", "from:"
        ]

        let lowercased = prefix.lowercased()

        // If just starting a filter, suggest filter types
        for filterPrefix in filterPrefixes {
            if filterPrefix.hasPrefix(lowercased) && filterPrefix != lowercased {
                return [filterPrefix]
            }
        }

        // If typing a filter value, could suggest common values
        // This would require access to the index to get popular values

        return []
    }
}
